File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.py
import math

def is_prime(n):
    if n < 2:
        return False
    
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

def solution(numbers):
    answer = 0
    visited = [False] * 7
    sarr = set()

    def dfs(nums, str, i):
        if i > len(nums):
            return None

        for i in range(len(nums)):
            if not visited[i]:
                visited[i] = True
                sarr.add(int(str + nums[i]))
                dfs(nums, str + nums[i], i +1)
                visited[i] = False

    dfs(numbers, "", 0)
    
    for _ in sarr:
        if is_prime(_): answer += 1

    return answer

File: 프로그래머스/2/42839. 소수 찾기/소수 찾기.java
import java.util.*;
class Solution {
    Set<Integer> set = new HashSet<>();
    public int solution(String numbers) {
        
        dfs("", numbers);
        int count = 0;
        for (int num : set) {
            if (isPrime(num)) {
                count++;
            }
        }
        return count;
    }
    
    public void dfs(String current, String remaining){
        if(!current.equals("")){
            set.add(Integer.parseInt(current));
        }
        
        for(int i=0; i<remaining.length(); i++){
            dfs(
                current + remaining.charAt(i),
                remaining.substring(0, i) + remaining.substring(i + 1)
            );
        }
    }
    
    public boolean isPrime(int n){
        if (n < 2) return false;
        for (int i = 2; i * i <= n; i++) {
            if (n % i == 0) return false;
        }
        return true;
    }
}